src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsWriter.cs

```diff
@@ -0,0 +1,187 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Buffers.Binary;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Data.SqlClient;
+
+namespace Microsoft.Data.SqlClientX.IO
+{
+    internal class TdsWriter
+    {
+        private readonly TdsWriteStream _writeStream;
+
+        // Header length constants
+        private const int OutputHeaderLen = TdsEnums.HEADER_LEN;
+
+        // Out buffer variables
+        private readonly byte[] _outBuff; // internal write buffer - initialize on login
+        private int _outBytesUsed = OutputHeaderLen; // number of bytes used in internal write buffer - initialize past header
+
+        /// <summary>
+        /// Instantiate TdsWriter with a TdsWriteStream
+        /// </summary>
+        /// <param name="stream">A Tds Writer stream to work with</param>
+        /// <exception cref="ArgumentNullException"></exception>
+        public TdsWriter(TdsWriteStream stream)
+        {
+            _writeStream = stream ?? throw new ArgumentNullException(nameof(stream));
```

Comment:
Instead of throwing an exception, please add a Debug.Assert. This basically means that a null stream just isn't allowed. There is nothing that a caller can do with the exception. Once the caller makes sure that a non-null stream is sent, then we dont need the null check.